spec/abci++/abci++_tmint_expected_behavior_002_draft.md

```diff
@@ -0,0 +1,184 @@
+>
+> First summarize the current text into: all good vs all bad
+> Then set out the rules (regex? grammar?)
+
+This section describes what the Application can expect from Tendermint
+
+The following section uses these definitions:
+
+* We define the _optimal case_ as a run in which (a) the system behaves synchronously, and (b) there are no Byzantine processes.
+  The optimal case captures the conditions that hold most of the time, but not always.
+* We define the _suboptimal case_ as a run in which (a) the system behaves asynchronously in round 0, height h -- messages may be delayed,
+  timeouts may be triggered --, (b) it behaves synchronously in all subsequent rounds (_r>0_), and (c) there are no Byzantine processes.
+  The _suboptimal case_ captures a possible glitch in the network, or some sudden, sporadic performance issue in some validator
+  (network card glitch, thrashing peak, etc.).
+* We define the _general case_ as a run in which (a) the system behaves asynchronously for a number of rounds,
+  (b) it behaves synchronously after some time, denoted _GST_, which is unknown to the processes,
+  and (c) there may be up to _f_ Byzantine processes.
+
+### `PrepareProposal`:  Application's expectations
+
+Given a block height _h_, process _p_'s Tendermint calls `RequestPrepareProposal` under
+the following conditions:
+
+* _p_'s Tendermint may decide at height _h_, without calling `RequestPrepareProposal`.
+  In the optimal case, this will happen often, as there will only be one proposer for height _h_
+  (the one for round 0).
+* _p_'s Tendermint may call `RequestPrepareProposal` with a block with no transactions, if
+  _ConsensusParams_ is configured to produce empty blocks when there are outstanding transactions.
+  If _ConsensusParams_ is configured to avoid empty blocks, any block passed with a call to
+  `RequestPrepareProposal` will contain at lesat one transaction.
+* In the optimal case, _p_'s Tendermint will call `RequestPrepareProposal` at most once,
+  as there is only one round.
+* In the suboptimal case, _p_'s Tendermint
+    * will not call `RequestPrepareProposal` for height _h_, if _p_ is neither the proposer
+      of round 0, nor round 1
+    * will call `RequestPrepareProposal` once for height _h_, if _p_ is either the proposer
+      of round 0, or round 1 (but not both).
+    * will call `RequestPrepareProposal` twice for height _h_, in the unlikely case that
+      _p_ is the proposer of round 0, and round 1.
+* In the general case, _p_'s Tendermint will potentially span many rounds. So, it will call
+  `RequestPrepareProposal` a number of times which is impossible to predict. Thus, the
+  Application will to return numerous _amended_ blocks via `ResponsePrepareProposal`.
+
+  If the application is fully executing the blocks it returns via `ResponsePrepareProposal`,
+  it should be careful about its usage of resources (memory, disk, etc.) as the number
+  of `PrepareProposal` for a particular height is not bounded.
```

Comment:
The same thing is effectively true for `ProcessProposal`. If an application is given multiple proposals via `ProcessProposal`, in both rounds `i` and `i+1` at a height, it is because Tendermint had no locked block when it received proposal `i+1`. It should therefore be fine to keep discarding old candidate state as new proposals arrive. Once it locks on a proposal, it no longer needs any of the old proposals.

Reply:
Thanks for sharing your thoughts. For info, when I wrote that text, I was thinking of "case 2" as defined in your comment.

Here are my answers:
1) This text is old. We will replace it with the grammar I showed last Fri, together with explanations. But I'd like to thank you for going through this and sharing your insights.
2) What you are proposing (i.e., keeping the latest candidate state the App has executed and dropping the previous ones) is a valid strategy to deal with the problem of _unbounded candidate states_. The text here, was merely **warning** the App developers they can't keep all candidate states, and that they need a strategy to _limit_ the candidate states they keep in memory.
   In the end, whatever strategy you choose to limit the candidate states you keep, your app must handle the (unlikely but possible) case when it has discarded the candidate state for the block passed in `FinalizeBlock`, and therefore, has to execute it again at that point.